power_grid_model_c/power_grid_model/include/power_grid_model/calculation_parameters.hpp

```diff
@@ -159,6 +159,17 @@ template <bool sym> struct MathModelParam {
     ComplexTensorVector<sym> source_param;
 };
 
+template <typename T>
+concept UpdateIsProgressive = requires(T t) { t.branch_param_to_change.size() > 0; };
```

Comment:
i don't think this will work. a `concept` is a `static` check, meaning that the compiler checks whether it compiles and does a compile-time verification and that's all it does.

in this particular case, it will check whether the semantics are valid, which it is, because `t.branch_param_to_size` exists and is an `std::vector<Idx>`, so `t.branch_param_to_change.size()` will compile and return an `std::integral` (by definition, `size_t`, and `size_t` and `decltype(0)` are comparable using `operator>`. since `t.branch_param_to_change.size() > 0;` will compile, this requirement is always true at compiletime, whenever you use `MathModelParamIncrement`.

If that check is indeed the check you want to have, it's better to just define the concept as either

```cpp
template <typename T>
concept increment_update = std::same_as<std::remove_cvref_t<T>, MathModelParamIncrement>;
```

Alternatively, if you want to check whether `t.branch_param_to_change` exists and is container-like, you can use

```cpp
template <typename T>
concept UpdateIsProgressive = requires(T t) {
    { t.branch_param_to_change } -> std::range;
    { *(t.branch_param_to_change.begin()) } -> Idx&;
};
```

If you don't want a compile-time check, please make it a function instead of a concept. you could even go as far as making both a compiletime and runtime check in the function:

```cpp
template <typename T>
concept math_model_param_c = std::same_as<std::remove_cvref_t<T>, MathModelParam> || std::same_as<std::remove_cvref_t<T>, MathModelParamIncrement>;

bool is_progressive_update(math_model_param_c auto const& param) {
    if constexpr (requires { param.branch_param_to_change }) {
        return param.branch_param_to_change.size() > 0;
    }
    return false;
}
```

of course, you'll also have to do something similar for the other attributes

Reply:
That's where I was confused about as well, [here](https://github.com/PowerGridModel/power-grid-model/pull/444#discussion_r1425096600). But I think it'll still work if my understanding is correct and that the function will be called by feeding it with the correct type of `MathModelParam`, namely `MathModelParamIncrement` that has property `::branch_param_to_change`, size zero or not, the logic should simply handle it. If of course it is not called in such a way and no compile time check is performed/necessary, I agree with you that this is not needed. With your comments above, I think the previous way (excluding the inheritance part) is fine already. Will revert to that.

Reply:
it works for the wrong reasons 😬 if you just want to know whether `branch_param_to_change` exists and has the correct type, just check for that. if you explicitly need `.size()`, check that its output result is `std::integral`

Reply:
😂 I don't disagree. It's gonna be removed anyways so maybe for the next `concept` 😉